From 041bd1a1fc737cd73b0b8a9f74909191a8acc9fe Mon Sep 17 00:00:00 2001
Message-Id: <041bd1a1fc737cd73b0b8a9f74909191a8acc9fe.1640253596.git.geliang.tang@suse.com>
From: Daniel Wagner <dwagner@suse.de>
Date: Wed, 1 Sep 2021 11:25:24 +0200
Subject: [PATCH] nvme: only call synchronize_srcu when clearing current path
Git-commit: 041bd1a1fc737cd73b0b8a9f74909191a8acc9fe
Patch-mainline: v5.15-rc1
References: bsn#51


The function nmve_mpath_clear_current_path returns true if the current
path has changed. In this case we have to wait for all concurrent
submissions to finish. But if we didn't change the current path, there
is no point in waiting for another RCU period to finish.

Signed-off-by: Daniel Wagner <dwagner@suse.de>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Geliang Tang <geliang.tang@suse.com>
---
 drivers/nvme/host/core.c |    9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

--- a/drivers/nvme/host/core.c
+++ b/drivers/nvme/host/core.c
@@ -3988,9 +3988,12 @@ static void nvme_ns_remove(struct nvme_n
 	list_del_rcu(&ns->siblings);
 	mutex_unlock(&ns->ctrl->subsys->lock);
 
-	synchronize_rcu(); /* guarantee not available in head->list */
-	nvme_mpath_clear_current_path(ns);
-	synchronize_srcu(&ns->head->srcu); /* wait for concurrent submissions */
+	/* guarantee not available in head->list */
+	synchronize_rcu();
+
+	/* wait for concurrent submissions */
+	if (nvme_mpath_clear_current_path(ns))
+		synchronize_srcu(&ns->head->srcu);
 
 	if (ns->disk->flags & GENHD_FL_UP) {
 		del_gendisk(ns->disk);
